### DISCUSSION TITLE
Boolean inputs get mapped to F32 when converting to `TypedModel`

### DISCUSSION THREAD

**Discussion by alexander-camuto** [sonos/tract#1161] [General]
Boolean inputs in ONNX get mapped to F32 ` TypedSource` nodes in tract. 

I was wondering if this was intentional behaviour. 

To replicate this you can run the tract CLI on the following graph composed solely of boolean inputs and AND | OR | XOR | EQUALS operations: 

[boolean.onnx.zip](https://github.com/sonos/tract/files/12419498/boolean.onnx.zip)

**Comment by kali** [comment]
```
TSAR 24/08 9:26 ~/dev/sonos/tract/issue-1161% cargo run -p tract  -- network.onnx dump --io-long                                                                                                   investigate-ci
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
     Running `/home/kali/dev/sonos/tract/target/debug/tract network.onnx dump --io-long`
┏ 0 Source input1
┃   * output fact #0: batch_size,3,Bool >2/0 >7/0 MODEL INPUT #1
┃┏ 1 Source input2
┃┃   * output fact #0: batch_size,3,Bool >2/1 >6/0 MODEL INPUT #2
┃┣┓
┣━━┓┃
┃┃┣┻ 2 And And_0
┃┃┃   * input fact  #0: 0/0> batch_size,3,Bool
┃┃┃     input fact  #1: 1/0> batch_size,3,Bool
┃┃┃   * output fact #0: batch_size,3,Bool >8/0  onnx::Equal_4
┃┃┃┏ 3 Source input3
┃┃┃┃   * output fact #0: batch_size,3,Bool >5/0 MODEL INPUT #3
┃┃┃┃┏ 4 Source input
┃┃┃┃┃   * output fact #0: batch_size,3,Bool >5/1 MODEL INPUT #0
┃┃┃┣┻ 5 Xor Xor_1
┃┃┃┃   * input fact  #0: 3/0> batch_size,3,Bool
┃┃┃┃     input fact  #1: 4/0> batch_size,3,Bool
┃┃┃┃   * output fact #0: batch_size,3,Bool >6/1  onnx::Or_5
┃┗┓┃
┃┃┣┻ 6 Or Or_2
┃┃┃   * input fact  #0: 1/0> batch_size,3,Bool
┃┃┃     input fact  #1: 5/0> batch_size,3,Bool
┃┃┃   * output fact #0: batch_size,3,Bool >7/1  onnx::And_6
┗┓┃
┃┣┻ 7 And And_3
┃┃   * input fact  #0: 0/0> batch_size,3,Bool
┃┃     input fact  #1: 6/0> batch_size,3,Bool
┃┃   * output fact #0: batch_size,3,Bool >8/1  onnx::Equal_7
┣┻ 8 Equals Equal_4
    * input fact  #0: 2/0> batch_size,3,Bool
      input fact  #1: 7/0> batch_size,3,Bool
    * output fact #0: batch_size,3,Bool  MODEL OUTPUT #0 output
```

All bools, no ?

**Comment by alexander-camuto** [comment]
hmmm that's correct. I've traced the issue back to a patch I added a few months ago such that this network could be analyzed: 

[network.onnx.zip](https://github.com/sonos/tract/files/12429503/network.onnx.zip)

![image](https://github.com/sonos/tract/assets/45801863/5ff3e2c7-ae85-4a88-83a2-b98cac4f996d)

I had to override the InferenceFact for tract to be able to analyse the graph in full (hence why everything gets overwritten with F32 types and explicitly specified tensor shapes). If I don't override this the analysis fails with: 

```bash
┏ 0 Source input
┃   ━━━ batch_size,3,5,5,F32
┣┻┻ 3 Resize /upsample/Resize
    ━━━ batch_size,3,10,10,F32
[2023-08-24T12:35:25.886682000Z ERROR tract] Error at stage analyse
    
    Caused by:
        0: Failed analyse for node #3 "/upsample/Resize" Resize
        1: Failed analyse for node #3 "/upsample/Resize" Resize
        2: Infering facts
        3: Applying rule Given2Rule { (inputs[0].shape, inputs[2]) }
        4: Undetermined symbol in expression: batch_size


```

I'm pretty sure this graph is correctly specified so this might be an issue ? not sure. 

It seems like because batch_size isn't determined at analysis time it is unable to apply the expansion along (dim=0), which is just 1 in this case. 

The pytorch graph that created this is nothing special either: 


```python
class Circuit(nn.Module):
    def __init__(self, inplace=False):
        super(Circuit, self).__init__()
        self.upsample = nn.Upsample(scale_factor=2, mode='nearest')

    def forward(self, x):
        y = self.upsample(x)
        return y




```

**Comment by kali** [comment]
I would not be surprised. Resize implementation is pretty weak and incomplete. I will have a look.

**Reply by alexander-camuto** [reply]
ty as always

**Reply by kali** [reply]
I think #1169 should work.

**Reply by alexander-camuto** [reply]
just tried it and it seems to :) thanks so much for the quick turnaround
